### PR TITLE
TEALv3: Min balance

### PR DESCRIPTION
**Pull request by ratik21**
- compute minimum balance
  - When creating or opting into an app, the minimum balance grows before the app code runs
  - When deleting or closing out, the minimum balance decreases after the app executes.
  - after increasing minBalance while creating asset/app, we should assert `account minimum balance < account.balance()`
  - updated few tests to assert minimum balance as well

- add `min_balance` opcode + tests

### CODE REVIEW

File: packages/runtime/src/ctx.ts
Comment (amityadav0) [inline packages/runtime/src/ctx.ts:152]: this is not quite clear, i think it is verifying if current balance is greater than minBalance, Please update the function to be more clear

File: packages/runtime/src/ctx.ts
Comment (amityadav0) [inline packages/runtime/src/ctx.ts:182]: i think this was discussed before, assets are not frozen for creator. right ?
Please add a link here about this assumption

File: packages/runtime/src/ctx.ts
Comment (amityadav0) [inline packages/runtime/src/ctx.ts:233]: why are we updating this ?

File: packages/runtime/src/ctx.ts
Comment (amityadav0) [inline packages/runtime/src/ctx.ts:201]:
```suggestion
    fromAccountAddr: AccountAddress, flags: SSCDeploymentFlags,
```

File: packages/runtime/src/ctx.ts
Comment (amityadav0) [inline packages/runtime/src/ctx.ts:192]: Params are incorrect

File: packages/runtime/src/ctx.ts
Comment (amityadav0) [inline packages/runtime/src/ctx.ts:250]:
```suggestion
  optInToApp (accountAddr: AccountAddress, appID: number): void {
```

File: packages/runtime/src/ctx.ts
Comment (amityadav0) [inline packages/runtime/src/ctx.ts:306]: Please add a link to docs here.

File: packages/runtime/src/ctx.ts
Comment (amityadav0) [inline packages/runtime/src/ctx.ts:304]: let's move `opt-out` from asset part to account class and reuse it here.

File: packages/runtime/test/integration/close-clear-ssc.ts
Comment (amityadav0) [inline packages/runtime/test/integration/close-clear-ssc.ts:101]:
```suggestion
```

File: packages/runtime/test/integration/close-clear-ssc.ts
Comment (amityadav0) [inline packages/runtime/test/integration/close-clear-ssc.ts:123]:
```suggestion
```

File: packages/runtime/src/ctx.ts
Comment (ratik21) [reply packages/runtime/src/ctx.ts:233]: i didn't. will check

File: packages/runtime/src/ctx.ts
Comment (ratik21) [reply packages/runtime/src/ctx.ts:152]: ok, let's create another function for this

File: packages/runtime/src/ctx.ts
Comment (ratik21) [reply packages/runtime/src/ctx.ts:182]: The concept of Opting In for creator does not make sense, because the create created the asset, and therefore it cannot opt-out of the ASA as well. 
While creating an asset,  asset holding is added in creator's account will all assets & default frozen false. Means that this code (`optInToASA`) will never be executed for asset creator.

I will add a link about ASA's in `account.addApp` where we create asset holding for creator

File: packages/runtime/src/account.ts
Comment (amityadav0) [inline packages/runtime/src/account.ts:224]: should we throw error if asset is not present in account ? or a warning ?

File: packages/runtime/src/ctx.ts
Comment (amityadav0) [inline packages/runtime/src/ctx.ts:187]:
```suggestion
```

File: packages/runtime/src/ctx.ts
Comment (amityadav0) [reply packages/runtime/src/ctx.ts:195]: `fromAccountAddr` is still missing

File: packages/runtime/src/ctx.ts
Comment (amityadav0) [inline packages/runtime/src/ctx.ts:196]: param is incorrect

File: packages/runtime/src/ctx.ts
Comment (amityadav0) [inline packages/runtime/src/ctx.ts:251]: doc string is incorrect, please update

File: packages/runtime/test/integration/close-clear-ssc.ts
Comment (amityadav0) [inline packages/runtime/test/integration/close-clear-ssc.ts:61]: please update it to check how much it is increased.

File: packages/runtime/test/integration/deleteApp.ts
Comment (amityadav0) [inline packages/runtime/test/integration/deleteApp.ts:71]: please add how much balance is increased, it is different for opt-in and creation. right ?

File: packages/runtime/test/src/interpreter/opcode-list.ts
Comment (amityadav0) [inline packages/runtime/test/src/interpreter/opcode-list.ts:4039]: i think this test should be in flow/integration tests, because here we check operations related to opcode-list only.
min_balance should push `min_balance`. this needs to be tested here

File: packages/runtime/test/src/interpreter/opcode-list.ts
Comment (amityadav0) [inline packages/runtime/test/src/interpreter/opcode-list.ts:4066]: :+1:

File: packages/runtime/test/src/runtime.ts
Comment (amityadav0) [inline packages/runtime/test/src/runtime.ts:213]: here we are testing this case, in opcode-list this is repeated

File: packages/runtime/src/account.ts
Comment (ratik21) [reply packages/runtime/src/account.ts:224]: before closing an asset(removing asset holding) we need to empty it's holding to the `closeRemTo` account (even if amount is zero). So we're verifying asset holding exists in `runtime.transferAsset` (before emptying balance). That's why i kept this code there only. Will add a comment here now.

File: packages/runtime/test/integration/close-clear-ssc.ts
Comment (ratik21) [reply packages/runtime/test/integration/close-clear-ssc.ts:61]: thought so initially, but then realized it doesn't make sense to compare constant increase (as we will be testing increment for exactly same value as we'll be increasing in code (..does that make sense?) ). Didn't do it here because we are testing min_balance in multiple flow (unlike ASA tests), which could be hard to interpret.
The important part is that minimum balance should grow after app/asset creation or opt-in, and reduce (to prev value) after opt-out/deletion.
If you want i can update to check exact value

File: packages/runtime/test/integration/deleteApp.ts
Comment (ratik21) [reply packages/runtime/test/integration/deleteApp.ts:71]: no, same `0.1ALGO`

File: packages/runtime/test/src/interpreter/opcode-list.ts
Comment (ratik21) [reply packages/runtime/test/src/interpreter/opcode-list.ts:4039]: > min_balance should push min_balance. this needs to be tested here

did that in previous test

> i think this test should be in flow/integration tests, because here we check operations related to opcode-list only.

in integration tests we test flow in smart contracts. Here we just add an asset and asset that `MinBalance` op should push raised balance.

File: packages/runtime/test/integration/close-clear-ssc.ts
Comment (amityadav0) [reply packages/runtime/test/integration/close-clear-ssc.ts:61]: agree this may not make sense now, but at some later point if storage costs are changed in code, then these tests will fail and we will know why

File: packages/runtime/test/src/runtime.ts
Comment (ratik21) [reply packages/runtime/test/src/runtime.ts:213]: no, in opcode list we specifically test execution of `MinBalance` opcode. There is a slight difference. Here we directly test account's min balance. I would prefer to keep it.

File: packages/runtime/test/integration/deleteApp.ts
Comment (amityadav0) [reply packages/runtime/test/integration/deleteApp.ts:71]: This one of for asa, I think for app it differs

File: packages/runtime/src/ctx.ts
Comment (ratik21) [reply packages/runtime/src/ctx.ts:194]: done

File: packages/runtime/test/src/runtime.ts
Comment (amityadav0) [reply packages/runtime/test/src/runtime.ts:213]: Ok, your call, better to have more tests

File: packages/runtime/test/integration/close-clear-ssc.ts
Comment (ratik21) [reply packages/runtime/test/integration/close-clear-ssc.ts:61]: makes sense

File: packages/runtime/test/integration/deleteApp.ts
Comment (ratik21) [reply packages/runtime/test/integration/deleteApp.ts:71]: oh yes, thought it was asa. 
For app, cost for creation & optin are different as in creation we increase balance using `global ints/bytes`, while in opt-in we increase/decrease balance referring to `local ints/bytes`. Base fee is same.

**Review by amityadav0 — APPROVED** [review]

**Comment by amityadav0** [comment]
Please update the change log before merging

**Pull request merged**
